Enhance graph edges

rbasu101 opened the pull request:
#2 Added new features for in-depth analysis of node-node relationships, such as:
- Selection of Nodes / Neighbors of Selected Node
- Retrieval of Edge Info between Nodes

rbasu101 committed:
graph.app.py:
    +Setup dash update to check for any 'selected edges' in plotly figure
    +Added new card panel to display textual info on selected edges
network_graph.py:
    +plotly figure now stores additional information on edges (approval,score,pmid,etc)
    +edges can now be selected
        -(Edge selection relies on hidden 'nodes' placed on the midpoint of edges to allow users to select them)

rbasu101 committed:
Enhance knowledge graph edges
Fixes #2

rbasu101 committed:
graph_app.py:
    +Added functionality to select nodes
    +Added functionality to show neighbor node list of a selected node
    +Added functioanlity to display edge-info
        +Can be done by selecting an 'edge node'
        +Can be done by selecting a node, and selecting a node from the node neighbor list
    +Added functioanlity to store networkx graph data in dash app for more detailed analysis of data
        -Optimizations for data storage within app are necessary (WIP).
network_graph.py:
    +Added neighbor list for each node in trace nodes
    +Provided functionality to generate json networkx graph

rbasu101 committed:
graph_app.py:
    -Added selected_node_display
network_graph.py:
    -Optimized initalize_network()
    -Reformatted create_trace_nodes()

rbasu101 committed:
graph_app.py:
    -Selecting an 'edge_node' clears neighbor list
    -Optimized update_edge_info()
        -Edge data is now read from netorkx graph json (instead of the plotly figure)
        -Caught error when selected_data is None
        -Minor change to boolean logic
network_graph.py:
    -Removed all 'customdata' storing edge data between nodes in plotly
        + The data is now retrived from the networkx graph json

mcannon068nw commented:
I found an error involving updating edge children when switching primary node:

![Screenshot 2023-10-24 at 2 10 36 PM](https://github.com/mcannon068nw/python-dgidb/assets/87494086/31bb183a-d2c5-4265-a3ba-d0b535662597)

I have only been able to reproduce it once but it seems to involve searching one gene, selecting an edge, then replacing the searched gene and trying to select a new edge.

mcannon068nw commented:
looks potentially like @app.callback (line 142 in graph_app.py) is not firing properly under certain conditions?

mcannon068nw commented on dgidb/graph_app.py:
```diff
@@ -46,26 +68,107 @@ def set_app_layout(app,plot,genes):
+        else:
+            return []
+
+def update_edge_info(app):
+    @app.callback(
+        Output('edge-info-text', 'children'),
+        [Input('selected-node', 'data'), Input('neighbor-dropdown', 'value')],
+        State('graph', 'data')
+    )
+    def update(selected_node,selected_neighbor,graph):
+        if selected_node is None:
+            return dash.no_update
+        if(selected_node['curveNumber'] == 1):
+            selected_data = get_node_data_from_id(graph['links'], selected_node['text'])
+            return "ID: " + str(selected_data['id']) + "\n\nApproval: " + str(selected_data['approval']) + "\n\nScore: " + str(selected_data['score']) + "\n\nAttributes: " + str(selected_data['attributes']) + "\n\nSource: " + str(selected_data['source']) + "\n\nPmid: " + str(selected_data['pmid'])
+        if selected_neighbor is not None:
+            edge_node_id = None
+            selected_node_is_gene = get_node_data_from_id(graph['nodes'], selected_node['text'])['isGene']
+            selected_neighbor_is_gene = get_node_data_from_id(graph['nodes'], selected_neighbor)['isGene']
+            if(selected_node_is_gene == selected_neighbor_is_gene):
+                return dash.no_update
+            elif(selected_node_is_gene):
+                edge_node_id = selected_node['text'] + " - " + selected_neighbor
+            elif(selected_neighbor_is_gene):
+                edge_node_id = selected_neighbor + " - " + selected_node['text']
+
+            selected_data = get_node_data_from_id(graph['links'], edge_node_id)
+            if selected_data is None:
+                return dash.no_update
+            return "ID: " + str(selected_data['id']) + "\n\nApproval: " + str(selected_data['approval']) + "\n\nScore: " + str(selected_data['score']) + "\n\nAttributes: " + str(selected_data['attributes']) + "\n\nSource: " + str(selected_data['source']) + "\n\nPmid: " + str(selected_data['pmid'])
+        return dash.no_update
```
I think the error posted in the comments happens because of something not properly firing within this block under certain conditions.

mcannon068nw reviewed (changes requested):
would like to look into the callback error a little bit to see if we can reproduce it before approving

rbasu101 committed:
graph_app.py:
    -User modification of gene dropdown resets selected node
        +Fixes Crash where a user selecting a new gene and selecting a new node from that gene causes a faulty query
    -Selecting a node always resets the selected neighbor
        +Fixes Crash where a user can sometimes create a query between a node and neighbor that are unrelated
    -Selecting a node always resets the edge-info
        +Ensures edge-info doesn't display information irrelevant to selected node

mcannon068nw reviewed (approved):
bug seems to be fixed, lgtm 👍
